[ENH] Vector-checking utilities

dPys opened the pull request:
Per Issue #24 
FR: Conformity check of bvecs and bvals (pure-python implementation)

# Todo:
1) reportlet viz and description of sampling scheme.
2) Parameterized smoke tests for each type of corrupted or incompatible bvec/bval file.

Comments/Suggestions welcome.
@dPys

arokem commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,334 @@
+    bvec_rescaled = [
+        b / np.linalg.norm(b) if not np.isclose(np.linalg.norm(b), 0) else b
+        for b in bvec
+    ]
```
```suggestion
    norms = np.linalg.norm(bvec, axis=-1)
    norms[np.isclose(norms, 0)] = 1 
    bvec_rescaled = bvec / norms[:, None]
```

arokem commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,334 @@
+        Amplitude square normed b-values array.
+    """
+    bval = np.sqrt(bval-bval.min())
+    bval_recaled = bval/bval.max() * 100
```
```suggestion
    bval_rescaled = bval/bval.max() * 100
```

arokem commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,334 @@
+
+def rescale_bval(bval):
+    """
+    Normalizes b-val by the square of the vector amplitude. If the
```
Is this really what this function does? I don't see the vectors being used.

dPys replied:
I rewrote this function, as it was unfinished when I first opened this PR.

arokem commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,334 @@
+    vec_array = np.row_stack(seen_vecs)
+    distances = np.linalg.norm(vec_array - bvec, axis=1)
+    distances_flip = np.linalg.norm(vec_array + bvec, axis=1)
+    return np.all(distances > cutoff) and np.all(distances_flip > cutoff)
```
I don't see `cutoff` being defined

dPys replied:
I eliminated this function altogether.

dPys commented:
Thanks for the quick review @arokem . Will make all requested changes and add tests by early next week! 

Cheers,
@dPys

dPys commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,334 @@
+    # Save bval/bvec text file equivalents in the case that these are the final
+    # versions of our gradient vectors
+    if final is True:
+        final_bval_path = "%s%s" % (vector_dir, '/final_bval.bval')
```
Since the function `make_gtab` gets reused each time the vectors are updated throughout the pipeline (this can happen at least 2-3 times), we can save the result with sequential integers as file suffices (this allows us to keep track of each change made to the vectors and in what order), but we can also use a `final` trigger that allows us to distinguish for the user the final version of the vectors to be used for post-preprocessing analysis. So by final here I am indeed referring to late stage of analysis.

oesteban replied:
I would allow nipype to keep track of this.

dPys commented:
@arokem -- basic tests are now included in dmriprep/utils/tests

dPys commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,334 @@
+    gtab_bvals = gtab.bvals.copy()
+    b0_thr_ixs = np.where(gtab_bvals < gtab.b0_threshold)[0]
+    gtab_bvals[b0_thr_ixs] = 0
+    gtab.b0s_mask = gtab_bvals == 0
```
Done!

oesteban commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,295 @@
+                a, b = (c, b) if os.path.exists(path_pattern % c) else (a, c)
+            return path_pattern % b
+        # Save gradient table to pickle
+        gtab_file = next_path(vector_dir + "/gtab_%s.pkl")
```
not sure I understand this. Anyways, why would you want to keep references to the original files?

dPys replied:
Provenance? To keep a record of each version of the gradient table that gets created. I can remove this though if you think it would be better handled through a logging mechanism (rather than through temporary files)?

oesteban replied:
As above, nipype should keep track of that for you.

oesteban commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,295 @@
+
+def rescale_bvec(bvec):
+    """
+    Normalizes b-vectors to be of unit length for the non-zero b-values. If the
```
This should be in imperative mood:

```suggestion
    Normalize b-vectors to be of unit length for the non-zero b-values.

    If the
```

oesteban commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,295 @@
+    -------
+    bvec_rescaled : m x n 2d array
+        Unit-normed b-vectors array.
+    """
```
```suggestion

    """
```
missing newline

oesteban commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,295 @@
+    # Enforce proper dimensions
+    bvec = bvec.T if bvec.shape[0] == 3 else bvec
+
+    # Normalize values not close to norm 1
```
I suggest a clearer algorithm:

```Python
axis = int(bvec.shape[0] != 3)  # Avoid reshaping data, just identify axis
b0s = np.linalg.norm(0) < B0_NORM_EPSILON  # Find indexes of b0
bvec[~b0s] = bvec[~b0s] / np.linalg.norm(bvec[~b0s], axis=axis)  # Normalize non-lowb
return bvec
```

If you don't want to modify in place, then, change the signature to have an `inplace=False` optional keyword argument and then add in the beginning:

```Python
if not inplace:
    bvec = bvec.copy()  # Do not modify in place
```

dPys replied:
Much cleaner. Thanks!

dPys replied:
Actually, shouldn't this be:
```
axis = int(bvec.shape[0] == 3)
b0s = np.linalg.norm(bvec) < B0_NORM_EPSILON
bvec[~b0s] = bvec[~b0s] / np.linalg.norm(bvec[~b0s], axis=axis)
```
?

oesteban commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,295 @@
+    return bvec_rescaled
+
+
+def rescale_bval(bval, bvec):
```
I think you want to merge both rescalings.

dPys replied:
Hmm, the only reason that I am hesitant on this one is that a user may not wish to rescale the bval. For instance, mrtrix allows the user to disable it altogether: See: https://mrtrix.readthedocs.io/en/latest/concepts/dw_scheme.html#b-value%20scaling

oesteban replied:
Please chime in here - https://github.com/bids-standard/bids-specification/pull/352#discussion_r337617411

But for dMRIPrep I would go with absolute b-vals without any doubt.

dPys replied:
Whatever our dmriprep developers agree is best, I'm happy to go with. Bear in mind that this is a separate correction from b-value rescaling performed based on b order of magnitude.

I am a bit concerned that like mrtrix outlines, mandatory b-val rescaling by vector amplitude, while beneficial in the vast majority of cases, could introduce "minor rounding errors in the components of the direction vectors" in some cases. 

If we do decide to make b-value vector rescaling optional, I could just add a `bval_scaling` flag, and make `True` the default?

dPys replied:
Also, I should note that any scaling of the b-vals gets rounded to the nearest integer (99/100 these are identical to the original). I'm still happy to disable b-val scaling altogether though if by 'absolute b-values' you meant the original, untouched?

oesteban commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,295 @@
+        slm = 'linear'
+        print('Fewer than 30 diffusion-weighted directions detected. To ensure adequate eddy current correction, eddy will be run using the --slm=linear option')
+    else:
+        slm = 'none'
```
I don't think this is necessary here, please run away of uber-functions trying to solve everything.

dPys replied:
Will do. If I transform checkvecs into an interface (it sounds like this may be the way to go), I could just add direction count and is_hemi as outputs? 

Would the slm logic here be best as a separate function to save for later, should it just be woven into the eddy interface somehow, or should it be incorporated in some other way? If we are planning to use FSL's eddy (even in the short-term), the slm arg will be important.

oesteban replied:
Please note before updating that I would not remove these functions, I think we want to wrap them up in Nipype interfaces.

Having these split functions makes it easier to implement unit tests (without all nipype's boilerplate).

oesteban commented:
BTW - happy to help as much as I can sending you PRs to your branch!

dPys commented:
Thanks so much @oesteban for going through this PR so carefully! Will make the requested changes and get back to you in a day or so.

@dPys

dPys commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,295 @@
+        final_bval_path = "%s%s" % (vector_dir, '/final_bval.bval')
+        final_bvec_path = "%s%s" % (vector_dir, '/final_bvec.bvec')
+        np.savetxt(final_bval_path, bvals.astype('int'), fmt='%d', newline=' ')
+        np.savetxt(final_bvec_path, bvecs.T.astype('float'), fmt='%8f')
```
Totally! Wasn't sure before if I should just go ahead and adopt that standard right away. But I like it and will outsource to a separate function.

dPys replied:
@oesteban -- One further question, should I even bother saving any of the corrected vector information in FSL-style bval/bvec files at any point (i.e. should I just save _everything_ that gets corrected to RASB .tsv)? I thought the plan was to also ensure backwards compatibility  + support for FSL's eddy or should that be left to FSL-related interfaces outside of the context of this initial vector-handling interface? @mattcieslak &  @arokem may also wish to chime in here on how exactly we want to deal with vector io for this particular interface. I'm totally open to what everyone thinks is best, I just want to make sure that I get it the way we want it the first time around.

dPys commented on dmriprep/utils/tests/test_vectors.py:
```diff
@@ -0,0 +1,84 @@
+
+
+def test_make_gtab():
+    """
```
Sounds good

oesteban commented on dmriprep/.idea/workspace.xml:
```diff
@@ -0,0 +1,360 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Please, do not resolve conversations until they are addressed :+1: - otherwise we will likely leave some requests out.

dPys replied:
Sounds good.

dPys commented:
@oesteban and @arokem --
The PR is now updated with most of the requested changes and then some. 
Additional changes:
*Function for writing the gtab out to .tsv.
*Tests now use sherbrooke_3shell with dipy's fetcher so that we don't have to include raw data in the package. 
*CheckVecs is now a nipype interface that lives in interfaces/vectors.py. 
*Significantly more modularity has been achieved so that separate routines live in separate functions.

Remaining to-do:
-Create a series of synthetic corrupted bval/bvecs to implement parametric pytesting. Where should these files be stored or should I write functions to create them on-the-fly?

oesteban commented on dmriprep/interfaces/vectors.py:
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""@authors: Derek Pisner, Oscar Estaban, Ariel Rokem"""
+from __future__ import division, print_function
```
no need for python 2 compatibility (left a comment somewhere else)

dPys replied:
yep, will remove!

oesteban commented on dmriprep/interfaces/vectors.py:
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""@authors: Derek Pisner, Oscar Estaban, Ariel Rokem"""
```
no need for authorship tags, should be covered by the contributing guidelines (that's why I was so looking forward to passing them)

dPys commented on dmriprep/interfaces/vectors.py:
```diff
@@ -0,0 +1,60 @@
+class CheckVecsOutputSpec(TraitedSpec):
+    """Output interface wrapper for CheckVecs"""
+    from dipy.core.gradients import GradientTable
+    gtab = traits.Instance(GradientTable)
```
Should we go 'all-in' and create the json sidecar now as well? The thing about RASB tsv alone is that we lose the b0 information that the gtab holds which we'll need to retain for other stages of preprocessing. The shared memory issue was exactly why I had been pickling the gtab initially, even if that wasn't a good long-term solution.

mattcieslak replied:
Wouldn't the B0 information just be a row of zeros?

dPys replied:
I'm referring to their _index_ and associated phase encoding directions.

dPys commented on dmriprep/interfaces/vectors.py:
```diff
@@ -0,0 +1,60 @@
+        setattr(self, '_pole', pole)
+        setattr(self, '_b0_ixs', b0_ixs)
+        setattr(self, '_gtab_tsv_file', gtab_tsv_file)
+        setattr(self, '_slm', slm)
```
will make this change!

dPys commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,433 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""@authors: Derek Pisner, Oscar Estaban, Ariel Rokem"""
+from __future__ import division, print_function
```
haha I keep forgetting that the days of backward compatibility pre-python3 are over (thank god)!

dPys commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,433 @@
+    # non-FSL form of eddy correction as well.
+    slm = set_slm(gtab, is_hemi)
+
+    return gtab, is_hemi, pole, b0_ixs, gtab_tsv_file, slm
```
I agree -- this would be nicely handled as a class. I like it, and will make that upgrade.

oesteban commented on dmriprep/interfaces/vectors.py:
```diff
@@ -0,0 +1,60 @@
+        setattr(self, '_is_hemi', is_hemi)
+        setattr(self, '_pole', pole)
+        setattr(self, '_b0_ixs', b0_ixs)
+        setattr(self, '_gtab_tsv_file', gtab_tsv_file)
```
I meant all of these can be replaced by:

```suggestion
self._results['gtab'] = gtab
self._results['is_hemi'] = is_hemi
self._results['pole'] = pole
self._results['b0_ixs'] = b0_ixs
self._results['gtab_tsv_file'] = gtab_tsv_file
```
and then the whole `list_outputs` method can be deleted.

dPys replied:
Got it!

dPys commented:
> Looking much much better, please keep up with the good work!

Glad to hear it.

oesteban commented:
one question to @arokem, @garikoitz, et al.:

I guess here we are also missing the conversion between image indices coordinates and RAS coordinates for the bvecs, is that correct?

I believe FSL and MRTrix use image coordinates - what is the case for dipy?

garikoitz commented:
I think FSL is image coordinates but mrtrix is scanner (real) coordinates

Wasn’t this part of Derek’s PR?

On Sun, 27 Oct 2019 at 16:00, Oscar Esteban <notifications@github.com>
wrote:

> one question to @arokem <https://github.com/arokem>, @garikoitz
> <https://github.com/garikoitz>, et al.:
>
> I guess here we are also missing the conversion between image indices
> coordinates and RAS coordinates for the bvecs, is that correct?
>
> I believe FSL and MRTrix use image coordinates - what is the case for dipy?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipreps/dmriprep/pull/26?email_source=notifications&email_token=ABCZAV2XEGFCVESD3HPV6XDQQYMSHA5CNFSM4JBQDZL2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECLKN6I#issuecomment-546744057>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCZAV3USWDXOX75U566AA3QQYMSHANCNFSM4JBQDZLQ>
> .
>

dPys commented:
> I think FSL is image coordinates but mrtrix is scanner (real) coordinates Wasn’t this part of Derek’s PR?
> […](#)
> On Sun, 27 Oct 2019 at 16:00, Oscar Esteban ***@***.***> wrote: one question to @arokem <https://github.com/arokem>, @garikoitz <https://github.com/garikoitz>, et al.: I guess here we are also missing the conversion between image indices coordinates and RAS coordinates for the bvecs, is that correct? I believe FSL and MRTrix use image coordinates - what is the case for dipy? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <#26?email_source=notifications&email_token=ABCZAV2XEGFCVESD3HPV6XDQQYMSHA5CNFSM4JBQDZL2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECLKN6I#issuecomment-546744057>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABCZAV3USWDXOX75U566AA3QQYMSHANCNFSM4JBQDZLQ> .

Well, I mean in my original PR, I had included the following:
```
def reorient_dwi(dwi_prep, bvecs, out_dir):
    """
    A function to reorient any dwi image and associated bvecs to RAS+.

    Parameters
    ----------
    dwi_prep : str
        File path to a dwi Nifti1Image.
    bvecs : str
        File path to corresponding bvecs file.
    out_dir : str
        Path to output directory.

    Returns
    -------
    out_fname : str
        File path to the reoriented dwi Nifti1Image.
    out_bvec_fname : str
        File path to corresponding reoriented bvecs file.
    """
     # Requires normalize_xform to be available in namespace
    fname = dwi_prep
    bvec_fname = bvecs
    out_bvec_fname = "%s%s" % (out_dir, '/bvecs_reor.bvec')

    input_img = nib.load(fname)
    input_axcodes = nib.aff2axcodes(input_img.affine)
    reoriented = nib.as_closest_canonical(input_img)
    normalized = normalize_xform(reoriented)
    # Is the input image oriented how we want?
    new_axcodes = ('R', 'A', 'S')
    if normalized is not input_img:
        out_fname = "%s%s%s%s" % (out_dir, '/', dwi_prep.split('/')[-1].split('.nii.gz')[0], '_reor_RAS.nii.gz')
        print("%s%s%s" % ('Reorienting ', dwi_prep, ' to RAS+...'))

        # Flip the bvecs
        input_orientation = nib.orientations.axcodes2ornt(input_axcodes)
        desired_orientation = nib.orientations.axcodes2ornt(new_axcodes)
        transform_orientation = nib.orientations.ornt_transform(input_orientation, desired_orientation)
        bvec_array = np.loadtxt(bvec_fname)
        if bvec_array.shape[0] != 3:
            bvec_array = bvec_array.T
        if not bvec_array.shape[0] == transform_orientation.shape[0]:
            raise ValueError("Unrecognized bvec format")
        output_array = np.zeros_like(bvec_array)
        for this_axnum, (axnum, flip) in enumerate(transform_orientation):
            output_array[this_axnum] = bvec_array[int(axnum)] * float(flip)
        np.savetxt(out_bvec_fname, output_array, fmt="%.8f ")
    else:
        out_fname = "%s%s%s%s" % (out_dir, '/', dwi_prep.split('/')[-1].split('.nii.gz')[0], '_noreor_RAS.nii.gz')
        out_bvec_fname = bvec_fname

    normalized.to_filename(out_fname)

    return out_fname, out_bvec_fname
```

mattcieslak commented:
I took a similar approach for reorienting images and bvecs (https://github.com/PennBBL/qsiprep/blob/master/qsiprep/interfaces/images.py#L358). If you're planning on using ants for most of the registration, then keeping things in LPS+ makes life a lot easier. Internally their world coordinate system is oriented this way and it makes operations like local rotations much easier. The downside is that if you're using smriprep there will be a lot of internal image axis flipping.

dPys commented:
> I took a similar approach for reorienting images and bvecs (https://github.com/PennBBL/qsiprep/blob/master/qsiprep/interfaces/images.py#L358). If you're planning on using ants for most of the registration, then keeping things in LPS+ makes life a lot easier. Internally their world coordinate system is oriented this way and it makes operations like local rotations much easier. The downside is that if you're using smriprep there will be a lot of internal image axis flipping.

@mattcieslak -- see if you like the updated version above. It now permits any of RAS/LAS/LPS as user-specified output orientations, rather than hard-coding it to any particular one. I do see what you mean about LPS being more ANTs friendly. Any ideas why the ANTs people haven't accepted RAS+ as a standard? As you mentioned, it makes things a bit trickier for us (plus maybe also risks confusion on the part of the user?) if we go with LPS and then have to keep flipping everything back...

dPys commented:
@oesteban --

I've made the requested changes plus some further enhancements:

1) I've created a class per your suggestion called VectorTools which now contains each of 8 different methods of io and data integrity checks. This should be much more concise than the original function that I was relying on before.
2) A DWI reorientation tool (that is also a method of VectorTools) to force the dwi and bvecs into a given image orientation with mutual correspondence (I default to RAS+ given our choice of a rasb tsv file, though see @mattcieslak 's comment above). 
3) I've started to add a number of synthetic corrupted bvecs/bvals for testing the integrity checkers. This part already raised the need for a tad more error handling, which I'll have to get to at some point soon.
4) I've fleshed out the core interface (now called CleanVecs), and added some switches for us to have better control over what the interface is doing each time we call it in workflows.
5) I've also added a few routines of my own for doing dwi/bval consistency checking -- basically all that's happening here is that I'm evaluating B0 locations from the _image_ using relative mean signal variation on the basis of 3 sd's above the mean across all 4d volumes. This essentially leverages the fact that the low B0 should always have dramatically higher SNR than diffusion-encoded volumes. I can then compare these indices with the indices specified by the bval (and found in the bvec rows) and ensure they correspond. Similarly, I go on to use sklearn's MeanShift to perform clustering of the mean signal intensities across image volumes to ensure that the number of clusters correspond to the number of unique _shells_ + B0's. This is a WIP (honestly surprised that I've never seen any other software do this), but I'm convinced that we should include something like it as a part of this vector-checking interface. It's also essentially instantaneous (~1 sec runtime total). Let me know what you think.
6) We still need to PR the gradient table rasb .tsv file io routines to nibabel per your suggestion. And we still need to come up with a json reader/writer for the gradient table metadata.

@dPys

oesteban commented:
> @dPys Basically, if the dwi image is in RAS, then we leave the bvec alone. If the image is in any other orientation, we reorient it to RAS and invert the corresponding axes of the bvecs that require flipping to get there. I was under the impression that pretty much all tools use image index coordinates as the convention? Is that not the case?

Why the DWIs must be in RAS+ coords? (poking @mattcieslak for this one, as he has more experience) I think how you access the data array is independent of the coordinates of the bvecs. If you have an interface that provides you both RASB and voxel coords (after all your checks), then nothing prevents you from grabbing the output you really need down the line.

> If so, and if we do plan to impose RAS+ orientation in dmriprep in general, it should definitely happen near the beginning of the workflow, and the bvec should be corrected accordingly...

I would try to avoid reorienting images by all means. If we still want to do it (and believe me, I cry when I find a NIfTI with an orientation other than RAS+), then that should be COMPLETELY independent of this PR. To calculate RASB you just need the affine of the image.

> @mattcieslak 
> I took a similar approach for reorienting images and bvecs (https://github.com/PennBBL/qsiprep/blob/master/qsiprep/interfaces/images.py#L358). If you're planning on using ants for most of the registration, then keeping things in LPS+ makes life a lot easier. Internally their world coordinate system is oriented this way and it makes operations like local rotations much easier. The downside is that if you're using smriprep there will be a lot of internal image axis flipping.

ITK/ANTs have a problem with NIfTIs define their orientation with the s-form matrix. In fMRIPrep we run a series of checks on the NIfTI headers, and make sure they all have q-form and s-form matrices that are consistent. Then you just need to be careful after any ITK-based processing as s-forms will not be updated (and thus, downstream steps will likely get orientation wrong).

This is potentially going to get much much better when we finish https://github.com/poldracklab/nitransforms (which btw is coming along pretty fast).

@dPys 
> * I've created a class per your suggestion called VectorTools which now contains each of 8 different methods of io and data integrity checks. This should be much more concise than the original function that I was relying on before.

I'll have a look ASAP

> * A vector reorientation tool (that is also a method of VectorTools) to force the dwi and bvecs into a given image orientation with mutual correspondence (I default to RAS+ given our choice of a rasb tsv file, though see @mattcieslak 's comment above).

Great! I'll give this all my love. I hope there are some nice tests there :D. Please grab from nitransfroms whatever you need, it's going to become a dependency anyways (better said, it will be part of nibabel).

> * I've started to add a number of synthetic corrupted bvecs/bvals for testing the integrity checkers. This part already raised the need for a tad more error handling, which I'll have to get to at some point soon.

Sounds good. I'll probably have to send a PR to your branch with stuff related to the above two points, so I'll have this in mind and try to help.

> * I've fleshed out the core interface (now called CleanVecs), and added some switches for us to have better control over what the interface is doing each time we call it in workflows.

:+1: 

> * I've also added a few routines of my own for doing dwi/bval consistency checking -- basically all that's happening here is that I'm evaluating B0 locations from the _image_ using relative mean signal variation on the basis of 3 sd's above the mean across all 4d volumes. This essentially leverages the fact that the low B0 should always have comparatively way higher SNR than gradient-encoded volumes. I can then compare these indices with the indices specified by the bval (and found in the bvec rows) and ensure they correspond. If they don't, then all that happens is a UserWarning gets raised. Similarly, I go on to use sklearn's MeanShift to perform clustering of the mean signal intensities across image volumes to ensure that the number of clusters correspond to the number of unique _shells_ + B0's. This is a WIP (honestly surprised that I've never seen any other software do this), but I'm convinced that we should include something like it as a part of this vector-checking interface. It's also essentially instantaneous (~1 sec runtime total). Let me know what you think.

I think this is awesome. If you want to split this to a subsequent PR, that will simplify your life to get this one in, simplify mine when reviewing and also reduce the risk we get out of focus in the PR (lengthening the review period).

> * We still need to PR the gradient table rasb .tsv file io routines to nibabel per your suggestion. 

Don't miss your sleep on this one. Let's test the waters first (WDYT, @effigies?)

> And we still need to come up with a json reader/writer for the gradient table metadata.

That's already there, and it's called pyBIDS :D.

dPys commented:
> > @dPys Basically, if the dwi image is in RAS, then we leave the bvec alone. If the image is in any other orientation, we reorient it to RAS and invert the corresponding axes of the bvecs that require flipping to get there. I was under the impression that pretty much all tools use image index coordinates as the convention? Is that not the case?
> 
> Why the DWIs must be in RAS+ coords? (poking @mattcieslak for this one, as he has more experience) I think how you access the data array is independent of the coordinates of the bvecs. If you have an interface that provides you both RASB and voxel coords (after all your checks), then nothing prevents you from grabbing the output you really need down the line.
> 
> > If so, and if we do plan to impose RAS+ orientation in dmriprep in general, it should definitely happen near the beginning of the workflow, and the bvec should be corrected accordingly...
> 
> I would try to avoid reorienting images by all means. If we still want to do it (and believe me, I cry when I find a NIfTI with an orientation other than RAS+), then that should be COMPLETELY independent of this PR. To calculate RASB you just need the affine of the image.
> 
> > @mattcieslak
> > I took a similar approach for reorienting images and bvecs (https://github.com/PennBBL/qsiprep/blob/master/qsiprep/interfaces/images.py#L358). If you're planning on using ants for most of the registration, then keeping things in LPS+ makes life a lot easier. Internally their world coordinate system is oriented this way and it makes operations like local rotations much easier. The downside is that if you're using smriprep there will be a lot of internal image axis flipping.
> 
> ITK/ANTs have a problem with NIfTIs define their orientation with the s-form matrix. In fMRIPrep we run a series of checks on the NIfTI headers, and make sure they all have q-form and s-form matrices that are consistent. Then you just need to be careful after any ITK-based processing as s-forms will not be updated (and thus, downstream steps will likely get orientation wrong).
> 
> This is potentially going to get much much better when we finish https://github.com/poldracklab/nitransforms (which btw is coming along pretty fast).
> 
> @dPys
> 
> > * I've created a class per your suggestion called VectorTools which now contains each of 8 different methods of io and data integrity checks. This should be much more concise than the original function that I was relying on before.
> 
> I'll have a look ASAP
> 
> > * A vector reorientation tool (that is also a method of VectorTools) to force the dwi and bvecs into a given image orientation with mutual correspondence (I default to RAS+ given our choice of a rasb tsv file, though see @mattcieslak 's comment above).
> 
> Great! I'll give this all my love. I hope there are some nice tests there :D. Please grab from nitransfroms whatever you need, it's going to become a dependency anyways (better said, it will be part of nibabel).
> 
> > * I've started to add a number of synthetic corrupted bvecs/bvals for testing the integrity checkers. This part already raised the need for a tad more error handling, which I'll have to get to at some point soon.
> 
> Sounds good. I'll probably have to send a PR to your branch with stuff related to the above two points, so I'll have this in mind and try to help.
> 
> > * I've fleshed out the core interface (now called CleanVecs), and added some switches for us to have better control over what the interface is doing each time we call it in workflows.
> 
> 👍
> 
> > * I've also added a few routines of my own for doing dwi/bval consistency checking -- basically all that's happening here is that I'm evaluating B0 locations from the _image_ using relative mean signal variation on the basis of 3 sd's above the mean across all 4d volumes. This essentially leverages the fact that the low B0 should always have comparatively way higher SNR than gradient-encoded volumes. I can then compare these indices with the indices specified by the bval (and found in the bvec rows) and ensure they correspond. If they don't, then all that happens is a UserWarning gets raised. Similarly, I go on to use sklearn's MeanShift to perform clustering of the mean signal intensities across image volumes to ensure that the number of clusters correspond to the number of unique _shells_ + B0's. This is a WIP (honestly surprised that I've never seen any other software do this), but I'm convinced that we should include something like it as a part of this vector-checking interface. It's also essentially instantaneous (~1 sec runtime total). Let me know what you think.
> 
> I think this is awesome. If you want to split this to a subsequent PR, that will simplify your life to get this one in, simplify mine when reviewing and also reduce the risk we get out of focus in the PR (lengthening the review period).
> 
> > * We still need to PR the gradient table rasb .tsv file io routines to nibabel per your suggestion.
> 
> Don't miss your sleep on this one. Let's test the waters first (WDYT, @effigies?)
> 
> > And we still need to come up with a json reader/writer for the gradient table metadata.
> 
> That's already there, and it's called pyBIDS :D.

effigies commented:
> > * We still need to PR the gradient table rasb .tsv file io routines to nibabel per your suggestion.
> 
> Don't miss your sleep on this one. Let's test the waters first (WDYT, @effigies?)

There are 83 comments in this thread. What's the specific proposal?

I can make a couple general comments, and if they don't fit what you're talking about, you can provide more context.

A pretty good model in the past has been to develop the IO routines you need in your own project, and once they're solidified, to add them to nibabel. For example, PySurfer did this with [`nibabel.freesurfer.io`](https://github.com/nipy/nibabel/blame/master/nibabel/freesurfer/io.py). This allows you to develop without being constrained by the nibabel release cycle.

Also, in case it's relevant, the data types have fairly heterogeneous APIs in nibabel. Images have one API, streamlines another, GIFTIs are barely related to NIfTIs, and the FreeSurfer IO functions work with dictionaries and numpy arrays, eschewing custom Python objects altogether. So don't feel obligated to contort yourself to mimic any of these APIs, if they're not natural. Figure out what makes sense for your data. (This also argues for developing locally and upstreaming later, as you can work through a few APIs if needed.)

dPys commented:
> > There are 83 comments in this thread. What's the specific proposal?
> 
> Sorry it was overwhelming - you clearly found your way to the proposal (i/o and storage of bvecs and bvals for DWI).
> 
> > I can make a couple general comments, and if they don't fit what you're talking about, you can provide more context.
> 
> Thanks, they do fit :)
> 
> > A pretty good model in the past has been to develop the IO routines you need in your own project, and once they're solidified, to add them to nibabel. For example, PySurfer did this with [`nibabel.freesurfer.io`](https://github.com/nipy/nibabel/blame/master/nibabel/freesurfer/io.py). This allows you to develop without being constrained by the nibabel release cycle.
> > Also, in case it's relevant, the data types have fairly heterogeneous APIs in nibabel. Images have one API, streamlines another, GIFTIs are barely related to NIfTIs, and the FreeSurfer IO functions work with dictionaries and numpy arrays, eschewing custom Python objects altogether. So don't feel obligated to contort yourself to mimic any of these APIs, if they're not natural. Figure out what makes sense for your data. (This also argues for developing locally and upstreaming later, as you can work through a few APIs if needed.)
> 
> Great, that confirms our approach. We will first develop something that works for us trying the code to be as nibabel-ish as possible but without that becoming a stopper. Then we can see if there's interest in adding them into nibabel.

@effigies -- as @oesteban noted, the specific proposal in this thread is to provide a single interface for all basic and advanced vector handling, which includes: storage/io, scaling, and integrity checking of the gradient table alongside the dwi. And yes, sorry if this thread's length feels overwhelming -- there are just many ways that these types of routines can be dealt with and I think it's good to have as much input from our contributors as possible on the standards we wish to set (so that this is done right the first time). Once vectors are adequately addressed with a versatile interface such as this, the rest of preprocessing becomes so much more straightforward.

effigies commented:
No worries on the length of the thread. That was mostly a statement that "I can't find this with `^Fnibabel`, and I'm not going to open all of the collapsed threads." Seems I hit the target, so it worked out for everybody.

> there are just many ways that these types of routines can be dealt with and I think it's good to have as much input from our contributors as possible on the standards we wish to set (so that this is done right the first time).

It's a good idea to get broad input, but getting an API right on the first try is not a huge deal, especially when you don't have any downstream projects importing your code. This is actually a good time to experiment, as you'll be working with these structures. The cost of changing is as low as it's ever going to be, so do something that works, and write useful abstractions as needed.

File formats and names are another matter. Mess with those after building any kind of user base, and you're going to have a bad time.

dPys commented:
> > @dPys Basically, if the dwi image is in RAS, then we leave the bvec alone. If the image is in any other orientation, we reorient it to RAS and invert the corresponding axes of the bvecs that require flipping to get there. I was under the impression that pretty much all tools use image index coordinates as the convention? Is that not the case?
> 
> Why the DWIs must be in RAS+ coords? (poking @mattcieslak for this one, as he has more experience) I think how you access the data array is independent of the coordinates of the bvecs. If you have an interface that provides you both RASB and voxel coords (after all your checks), then nothing prevents you from grabbing the output you really need down the line.
> 
> > If so, and if we do plan to impose RAS+ orientation in dmriprep in general, it should definitely happen near the beginning of the workflow, and the bvec should be corrected accordingly...
> 
> I would try to avoid reorienting images by all means. If we still want to do it (and believe me, I cry when I find a NIfTI with an orientation other than RAS+), then that should be COMPLETELY independent of this PR. To calculate RASB you just need the affine of the image.
> 
> > @mattcieslak
> > I took a similar approach for reorienting images and bvecs (https://github.com/PennBBL/qsiprep/blob/master/qsiprep/interfaces/images.py#L358). If you're planning on using ants for most of the registration, then keeping things in LPS+ makes life a lot easier. Internally their world coordinate system is oriented this way and it makes operations like local rotations much easier. The downside is that if you're using smriprep there will be a lot of internal image axis flipping.
> 
> ITK/ANTs have a problem with NIfTIs define their orientation with the s-form matrix. In fMRIPrep we run a series of checks on the NIfTI headers, and make sure they all have q-form and s-form matrices that are consistent. Then you just need to be careful after any ITK-based processing as s-forms will not be updated (and thus, downstream steps will likely get orientation wrong).
> 
> This is potentially going to get much much better when we finish https://github.com/poldracklab/nitransforms (which btw is coming along pretty fast).

@oesteban -- As a quick note of clarification (and @mattcieslak please still chime in on this one as well with your take), I did not mean to suggest that the DWI's need to be reoriented, and I will gladly remove those lines from this particular PR to expedite review and allow for further brainstorming. However, I do think that it is critical that there be careful and ongoing correspondence between the orientation of the dwi's and the bvecs, even if at the end of the day that relationship is simply represented in metadata with accompanying transforms. In the past I've defaulted to simply enforcing RAS of both the image and bvecs just to make this way easier, but I'm totally on board with not reorienting the image if the preference is to avoid it. Down the road, we will just need to upgrade the rasb tsv reading writing in this PR so that it can also perform the necessary column inversions (i.e. in the case that the image is _not_ in RAS) before it is saved.
> 
> > * I've created a class per your suggestion called VectorTools which now contains each of 8 different methods of io and data integrity checks. This should be much more concise than the original function that I was relying on before.
> 
> I'll have a look ASAP
> 
> > * A vector reorientation tool (that is also a method of VectorTools) to force the dwi and bvecs into a given image orientation with mutual correspondence (I default to RAS+ given our choice of a rasb tsv file, though see @mattcieslak 's comment above).
> 
> Great! I'll give this all my love. I hope there are some nice tests there :D. Please grab from nitransfroms whatever you need, it's going to become a dependency anyways (better said, it will be part of nibabel).
> 
> > * I've started to add a number of synthetic corrupted bvecs/bvals for testing the integrity checkers. This part already raised the need for a tad more error handling, which I'll have to get to at some point soon.
> 
> Sounds good. I'll probably have to send a PR to your branch with stuff related to the above two points, so I'll have this in mind and try to help.
> 
> > * I've fleshed out the core interface (now called CleanVecs), and added some switches for us to have better control over what the interface is doing each time we call it in workflows.
> 
> 👍
> 
> > * I've also added a few routines of my own for doing dwi/bval consistency checking -- basically all that's happening here is that I'm evaluating B0 locations from the _image_ using relative mean signal variation on the basis of 3 sd's above the mean across all 4d volumes. This essentially leverages the fact that the low B0 should always have comparatively way higher SNR than gradient-encoded volumes. I can then compare these indices with the indices specified by the bval (and found in the bvec rows) and ensure they correspond. If they don't, then all that happens is a UserWarning gets raised. Similarly, I go on to use sklearn's MeanShift to perform clustering of the mean signal intensities across image volumes to ensure that the number of clusters correspond to the number of unique _shells_ + B0's. This is a WIP (honestly surprised that I've never seen any other software do this), but I'm convinced that we should include something like it as a part of this vector-checking interface. It's also essentially instantaneous (~1 sec runtime total). Let me know what you think.
> 
> I think this is awesome. If you want to split this to a subsequent PR, that will simplify your life to get this one in, simplify mine when reviewing and also reduce the risk we get out of focus in the PR (lengthening the review period).

Cool. I will move this WIP to a separate PR then.

> 
> > * We still need to PR the gradient table rasb .tsv file io routines to nibabel per your suggestion.
> 
> Don't miss your sleep on this one. Let's test the waters first (WDYT, @effigies?)
> 
> > And we still need to come up with a json reader/writer for the gradient table metadata.
> 
> That's already there, and it's called pyBIDS :D.
Yep, we may need to PR directly to pyBIDS then to cover the additional metadata needed per the spec that we're deciding on in [#349](https://github.com/bids-standard/bids-specification/issues/349)

dPys commented:
> No worries on the length of the thread. That was mostly a statement that "I can't find this with `^Fnibabel`, and I'm not going to open all of the collapsed threads." Seems I hit the target, so it worked out for everybody.
> 
> > there are just many ways that these types of routines can be dealt with and I think it's good to have as much input from our contributors as possible on the standards we wish to set (so that this is done right the first time).
> 
> It's a good idea to get broad input, but getting an API right on the first try is not a huge deal, especially when you don't have any downstream projects importing your code. This is actually a good time to experiment, as you'll be working with these structures. The cost of changing is as low as it's ever going to be, so do something that works, and write useful abstractions as needed.
> 
> File formats and names are another matter. Mess with those after building any kind of user base, and you're going to have a bad time.

Totally!

dPys commented:
Alright @oesteban -- I've made the requested changes. 

*Removed image reorientation and instead added basic vector inversion routines to VectorTools that force bvecs to RAS when saving gtab to .tsv (i.e. if image is not in RAS and thus neither is the bvecs)
*Removed image_gradient_consistency_check function in order to create a separate PR for it.
*Updated tests and confirmed corruption testing examples with pytest.raises(Exception)

Please let me know if there is anything else that remains for this PR. Otherwise, we should be pretty close to merging?

oesteban commented:
> dwi image gets reoriented, then the bvecs need to be inverted in the corresponding column(s) to that reorientation, right?

If the DWI are correctly reoriented, the voxel centers should be in the same RAS+ (physical coordinates). If the bvecs are in RAS+ orientation, the answer should be no.

> So if our plan is to save the vectors in rasb tsv, but the raw dwi image was acquired in LAS for instance, then the rasb tsv reader/writer will also have to perform the necessary column inversions to get the vectors to RAS before the gradient table is saved, no?

As I said, if the image is correctly reoriented, then sampling on any given (x, y, z) coordinates in mm in the LAS and RAS oriented versions should give you exactly the same value. So, RAS+ coordinates are absolute.

> Yep, is the plan then to PR directly to pyBIDS to cover the additional metadata needed per the spec that we're deciding on in [#349](https://github.com/bids-standard/bids-specification/issues/349)?

No need to send any patches. PyBIDS can do this already.

mattcieslak commented:
In my experience DWIs and bvecs never come out of dcm2niix in RAS+. They are always in LAS+ (the FSL-standard "RPI"). The reason behind this is that anything other than LAS+ will break the TOPUP/eddy integration. Early in qsiprep I converted everything to LPS+ so that voxel coordinates and world coordinates would have the same sign in ANTs and DSI Studio. However, we discovered that if you supply LPS+ data to TOPUP it will calculate a fieldmap that is flipped along the j axis and eddy will apply this warp to its outputs, resulting in horrible new distortions (https://github.com/PennBBL/qsiprep/issues/35). Side note: eddy will also misinterpret fieldmaps calculated by sdcflows.

Now qsiprep decides which initial orientation the pipeline will use based on which algorithm the user selects for motion correction. LAS+ for TOPUP/eddy and LPS+ for our python/ANTs-based method. This is needless complexity and is very frustrating. On the bright side it helped us figure out this fieldmap mess, which we would have missed if we went with RAS+.

Until there is a widely-accepted alternative to TOPUP/eddy you will be stuck with LAS+, at least for a portion of the workflow.

dPys commented:
> In my experience DWIs and bvecs never come out of dcm2niix in RAS+. They are always in LAS+ (the FSL-standard "RPI"). The reason behind this is that anything other than LAS+ will break the TOPUP/eddy integration. Early in qsiprep I converted everything to LPS+ so that voxel coordinates and world coordinates would have the same sign in ANTs and DSI Studio. However, we discovered that if you supply LPS+ data to TOPUP it will calculate a fieldmap that is flipped along the j axis and eddy will apply this warp to its outputs, resulting in horrible new distortions ([PennBBL/qsiprep#35](https://github.com/PennBBL/qsiprep/issues/35)). Side note: eddy will also misinterpret fieldmaps calculated by sdcflows.
> 
Yeah I noticed dcm2niix likes to output to LAS+ too. The point about supplying an LPS oriented image to topup leading to weird eddy outputs is unsurprising.

> Now qsiprep decides which initial orientation the pipeline will use based on which algorithm the user selects for motion correction. LAS+ for TOPUP/eddy and LPS+ for our python/ANTs-based method. This is needless complexity and is very frustrating. On the bright side it helped us figure out this fieldmap mess, which we would have missed if we went with RAS+.
> 

This is interesting. So it sounds like qsiprep does enforce reorientation of the dwi image then?
It sounds like with dmriprep, we are trying to find a way to avoid that entirely.

> Until there is a widely-accepted alternative to TOPUP/eddy you will be stuck with LAS+, at least for a portion of the workflow.

I can say that in the past when I've reoriented the raw image and corresponding bvecs to RAS, I never ran into any trouble with topup/eddy so I'm surprised by this. That said, I now want to do a bit of testing so that I can see if I just missed something outright.

oesteban commented:
Hey Derek, I'm actively working on your PR. Expect a PR to your branch by Saturday.

oesteban commented:
Okay, can you resync with master? There are two alternatives. Before starting PLEASE MAKE SURE AND CHECK SEVERAL TIMES WHETHER YOUR UPSTREAM REMOTE POINTS TO nipreps/dmriprep. :D


The pro alternative:
```
git fetch upstream
git rebase upstream/master
<fix potential conflicts>
```

The okay alternative:
```
git fetch upstream
git merge upstream/master
```

dPys commented:
Rebased, and should be good to go now! Apologies if you ran into some some merge conflicts along the way.

oesteban commented:
Please check and merge https://github.com/dPys/dmriprep/pull/1 first.

Basically, I worked your code to improve in several fronts.

Also, for this to be merged we need to have tests passing. Please check the pytest command I left in one comment.

We'll need to go over your tests, but I can't do that before Monday

pep8speaks commented:
Hello @dPys, Thank you for updating!

* In the file [`dmriprep/interfaces/vectors.py`](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/interfaces/vectors.py):

> [Line 101:17](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/interfaces/vectors.py#L101): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`dmriprep/utils/tests/test_vectors.py`](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/tests/test_vectors.py):

> [Line 20:5](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/tests/test_vectors.py#L20): [F841](https://duckduckgo.com/?q=pep8%20F841) local variable 'fdwi' is assigned to but never used
> [Line 29:35](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/tests/test_vectors.py#L29): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 71:100](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/tests/test_vectors.py#L71): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 99 characters)
> [Line 72:96](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/tests/test_vectors.py#L72): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 72:100](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/tests/test_vectors.py#L72): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (118 > 99 characters)
> [Line 74:100](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/tests/test_vectors.py#L74): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 99 characters)
> [Line 75:100](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/tests/test_vectors.py#L75): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 99 characters)
> [Line 104:38](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/tests/test_vectors.py#L104): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 106:62](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/tests/test_vectors.py#L106): [E712](https://duckduckgo.com/?q=pep8%20E712) comparison to True should be 'if cond is True:' or 'if cond:'
> [Line 108:64](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/tests/test_vectors.py#L108): [E712](https://duckduckgo.com/?q=pep8%20E712) comparison to True should be 'if cond is True:' or 'if cond:'
> [Line 118:46](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/tests/test_vectors.py#L118): [F841](https://duckduckgo.com/?q=pep8%20F841) local variable 'e_info' is assigned to but never used
> [Line 142:100](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/tests/test_vectors.py#L142): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 99 characters)
> [Line 144:100](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/tests/test_vectors.py#L144): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (113 > 99 characters)

* In the file [`dmriprep/utils/vectors.py`](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/vectors.py):

> [Line 341:25](https://github.com/nipreps/dmriprep/blob/175fa706991e3d67bbb4c40d1e778037605760ee/dmriprep/utils/vectors.py#L341): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

To test for issues locally, `pip install flake8` and then run `flake8 dmriprep`.

oesteban commented on dmriprep/utils/tests/test_vectors.py:
```diff
@@ -0,0 +1,192 @@
+    from os.path import expanduser, join
+    from dipy.core.gradients import GradientTable
+
+    fetch_sherbrooke_3shell()
```
This will hit dipy servers (whatever CDN they are using) for every test. We can move it to `dmriprep/conftest.py` and run it only once. Deal?

Then you make the files via a fixture (I'm sure you'll love fixtures of pytest).

dPys replied:
Deal!

oesteban commented on dmriprep/utils/tests/test_vectors.py:
```diff
@@ -0,0 +1,192 @@
+    assert os.path.isfile(gtab_tsv_file) is True
+    assert np.isclose(bvecs_read.all(), bvecs.all())
+    assert np.isclose(bvals_read.all(), bvals.all())
+    os.remove(gtab_tsv_file)
```
Too much boilerplate. I'll send a PR showing you how to use the `tmpdir` fixture so that tests run in temporal folders.

dPys replied:
> I think this is almost there, only the test file remaining. I'll write to your branch directly this time since I believe you won't mind and it will be much faster for iterating.
> 
> Then, I'll squash some commits before asking others to review (finalizing the whole work).
> 
> This has been a bit painful process (especially for you - you've done a tone of work). I hope it's been as illustrative as painful too.
> 
> We want the process to be like a skeet shooting game. When we open an issue, that's shouting for another clay pigeon target. Sending a PR is actually shooting. Tests and reviewers are the referees determining whether we hit the target.
> 
> The earlier version of the PR attempted to hit several targets at once. That's harder for the shooter, and for the referees as well. It also requires more tests. The more focused a PR is, the better. Perhaps, I'm a bit stubborn with this because I actually have fallen for this plenty of times (and sometimes I still do!).
> 
> It is tempting to try to solve everything at once but in the long that usually backfires.
> 
> I'll be super-happy to see this work merged in master!

Yeah but I think to some extent for this PR, this volume of work was perhaps a bit inevitable (and the boundaries harder to define) since we are essentially setting the spec and io all at once that will provide the foundation for vector handling throughout the workflow. That is, once we get this the way we want it (hopefully close), this code will get reused again and again, and in different ways, each time a change is made to the dwi. Vector handling is just super broad and a lot of information/switches are unavoidably tethered to it. I think making more piecewise PR's down the road will be wayyy easier. (e.g. eddy only, denoising only, etc.)

oesteban commented on dmriprep/utils/tests/test_vectors.py:
```diff
@@ -0,0 +1,192 @@
+    assert slm == 'none'
+
+
+def test_clean_vecs():
```
this test is misplaced. should go under the interfaces directory.

dPys replied:
Yep!

oesteban commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,341 @@
+
+    @property
+    def bvecs(self):
+        return self._bvecs
```
I'll add docstrings to these getters where they are missing.

dPys commented:
@oesteban -- going through your edits now, they look terrific! Thanks again for trimming this down. It's also been a nice opportunity for me to get a good sense of the stylistic elements that you're looking for in dmriprep.

I resolved some merge conflicts and corrected a few missing pieces of docstring, but will be rebasing this PR shortly.

@dPys

dPys commented:
Also, I noticed that you pruned out read_gradients_tsv and write_gradients_tsv.  It seems like they are still useful functions that would be better directed towards nibabel. Is that the plan for them?

def read_gradients_tsv(rasb_tsv_file):
    """
    Read bvals/bvecs from a RAS-B tsv file.
    Parameters
    ----------
    rasb_tsv_file : pathlike or str
        Path to .tsv file containing RAS-B gradient table.
    Returns
    -------
    bvals : 1d array
        Raw b-values float array.
    bvecs : m x n 2d array
        Raw b-vectors array.
    """
    RASB_table = np.genfromtxt(rasb_tsv_file, delimiter="\t")
    bvecs = RASB_table[1:, 0:3]
    bvals = RASB_table[1:, 3]
    return bvals, bvecs


def write_gradients_tsv(bvecs, bvals, out_file):
    """
    Save bvals/bvecs as RAS-B tsv.
    Parameters
    ----------
    bvals : 1d array
        Raw b-values float array.
    bvecs : m x n 2d array
        Raw b-vectors array.
    out_file : pathlike or str
        Path to .tsv file containing RAS-B gradient table.
    """
    RASB_table = np.column_stack([bvecs, bvals])
    with open(out_file, 'wb') as f:
        f.write(b'R\tA\tS\tB\n')
        np.savetxt(f, RASB_table.astype('float'), fmt='%.6f', delimiter="\t")
    f.close()
    return

oesteban commented:
Please, please, please, please, hold on sending more commits while I rewrite the history...

oesteban commented:
> Is that the plan for them?

Both are done when the rasb_file is a string or a path - both read/write functionalities are there.

josephmje commented:
Just a question from my end. Is the idea to use the output rasb tsv file instead of the input bvec and bval files for all downstream operations?

dPys commented:
> Just a question from my end. Is the idea to use the output rasb tsv file instead of the input bvec and bval files for all downstream operations?

@josephmje -- I think as much as possible, yes. It is simply a more concise way to storing the vector information so that: 1) we are not restricted to relying on two separate files; and 2) we always know the orientation of the vectors relative to RAS (information that is absent from traditional bvecs alone). Now, should we choose to include FSL-based tools like EDDY (at least until dmriprep develops better ones), we would still need to augment interfacing to handle rasbtsv since those tools fundamentally rely on the bvec/bval formats -- but this shouldn't be too big a deal.

oesteban commented:
As it is implemented in #26 right now, you will have access to both formats at all times (RAS+B and FSL-style), so you just connect the right input to the interface.

dPys commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,359 @@
+            np.savetxt(str(path), self.gradients,
+                       delimiter='\t', header='\t'.join('RASB'),
+                       fmt=['%.8f'] * 3 + ['%g'])
+        if bvecs:
```
Do we want to include a conditional on the transpose in case the bvecs is already fsl-style? Not sure if the interfaces enforces this already, but figured it would be worth confirming...
```
            if bvecs.shape[-1] == 3:
                np.savetxt(str(bvecs), self.bvecs.T, fmt='%.6f')
            else:
                np.savetxt(str(bvecs), self.bvecs, fmt='%.6f')
```

dPys commented on dmriprep/conftest.py:
```diff
@@ -5,6 +5,28 @@
+}
 
 
 @pytest.fixture(autouse=True)
```
This is awesome.

dPys commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,359 @@
+            np.savetxt(str(path), self.gradients,
+                       delimiter='\t', header='\t'.join('RASB'),
+                       fmt=['%.8f'] * 3 + ['%g'])
+        if bvecs:
```
Do we need to include a conditional here or does the shape get enforced by the interface? hard to tell
```
            if bvecs.shape[-1] == 3:
                np.savetxt(str(bvecs), self.bvecs.T, fmt='%.6f')
            else:
                np.savetxt(str(bvecs), self.bvecs, fmt='%.6f')
```

oesteban replied:
At this point, `self.bvecs` has (Nx3) dimensions. Please note that `bvecs` here just contains a string or a Path (perhaps improving the docstring here would be nice).

dPys replied:
Oh, I see that now. I wonder if to avoid confusion, we should rename bvecs to fbvecs for every place we're referring to a file string/Path, and go with bvecs any time we're referring to an array?

oesteban replied:
The function is called `to_filename`, I don't think changing the name of the arguments will change much. I guess it is more of an interface problem, i.e., should we have `to_filename` store only RASB and then add two more methods like `bvecs_to_filename` and `bvals_to_filename` or a `write_bvec_bval(path)` that adds `.bval` and `.bvec`.

Another alternative is:

``to_filename(filename, filetype='rasb')`` and then switch to bvec/bval mode if filetype is `'fsl'` I would probably like this last one better.

dPys replied:
Fair enough. Well I like the latter alternative best, but yes I think we need something like a write_bvec_bval(path) here. I sort of had something like this previously in the interface with:
```
if self.inputs.save_fsl_style is True:
    fbval_out_file, fbvec_out_file = vt.save_vecs_fsl()
```
but the to_filename would be best

oesteban replied:
Implemented the `to_filename(filename, filetype='fsl')` route. :+1:

josephmje commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,359 @@
+
+    def b0mask(self):
+        """Get a mask of low-b frames."""
+        return np.squeeze(self.gradients[..., -1] > self._b0_thres)
```
Should this be less than the threshold?

oesteban replied:
Indeed! Thanks for noticing!

josephmje commented on dmriprep/utils/vectors.py:
```diff
@@ -0,0 +1,359 @@
+            raise ValueError('The number of b-vectors and b-values do not match')
+        self._bvals = np.array(value)
+
+    def b0mask(self):
```
```suggestion
    @property
    def b0mask(self):
```

oesteban commented:
Okay, i don't think it'd be premature to merge. Let's do this!

oesteban commented:
@dpys, please send a PR adding your name to the zenodo file. Please remember to include `[skip ci]` in the commit title (the first line)
